elasticdl/python/allreduce/tensorflow_optimizer.py

```diff
@@ -0,0 +1,590 @@
+# Copyright 2020 The ElasticDL Authors. All rights reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+import horovod.tensorflow as hvd
+import tensorflow as tf
+from horovod.tensorflow import _LegacyOptimizer
+
+def complement_value_from_env_if_none(original_value, key, clz):
+    if original_value is not None:
+        return original_value
+
+    if key in os.environ:
+        return clz(os.environ.get(key))
+
+    raise ValueError("Cannot complement the value with key {} from environment".format(key))
+
+class AdjustBackwardPassesPerStepHook(tf.train.SessionRunHook):
+    """
+    Hooks that adjusts `backward_passer_per_step` according to
+    the horovod size dynamically.
+    """
+
+    def __init__(
+        self, backward_passes_per_step_var, global_batch_count_per_step=None
```

Comment:
`backward_passes_per_step_var` -> `backward_passes_per_step`?

Reply:
We can rename it because we have already indicate that this parameter should be a tf.Variable.